File: app/src/main/java/com/github/goldy1992/mp3player/client/callbacks/playback/MyPlaybackStateCallback.java
package com.github.goldy1992.mp3player.client.callbacks.playback;

import android.os.Handler;
import android.support.v4.media.session.PlaybackStateCompat;

import com.github.goldy1992.mp3player.client.callbacks.AsyncCallback;

import java.util.HashSet;
import java.util.Set;

import javax.inject.Inject;
import javax.inject.Named;

public class MyPlaybackStateCallback extends AsyncCallback<PlaybackStateCompat> {
    private static final String LOG_TAG = "MY_PLYBK_ST_CLLBK";
    private final Set<PlaybackStateListener> listeners;

    @Inject
    public MyPlaybackStateCallback(@Named("main") Handler handler) {
        super(handler);
        this.listeners = new HashSet<>();
    }

    @Override
    public void processCallback(PlaybackStateCompat data) {
        for (PlaybackStateListener listener : listeners) {
            listener.onPlaybackStateChanged(data);
        }
    }

    public synchronized void registerPlaybackStateListener(PlaybackStateListener listener) {
        listeners.add(listener);
    }
    /**
     * @param listener the PlaybackStateListener to be removed
     * @return true if the listener was removed successfully
     */
    public synchronized boolean removePlaybackStateListener(PlaybackStateListener listener) {
        return listeners.remove(listener);
    }

    public Set<PlaybackStateListener> getListeners() {
        return listeners;
    }
}

File: client/src/main/java/com/github/goldy1992/mp3player/client/callbacks/playback/MyPlaybackStateCallback.kt
package com.github.goldy1992.mp3player.client.callbacks.playback

import android.support.v4.media.session.PlaybackStateCompat
import com.github.goldy1992.mp3player.commons.LogTagger
import javax.inject.Inject

class MyPlaybackStateCallback

    @Inject
    constructor() : LogTagger {

    private val listeners: MutableSet<PlaybackStateListener> = HashSet()

    fun processCallback(data: PlaybackStateCompat) {
        for (listener in listeners) {
            listener.onPlaybackStateChanged(data)
        }
    }


    fun registerPlaybackStateListener(listener: PlaybackStateListener) {
        listeners.add(listener)
    }

    /**
     * @param listener the PlaybackStateListener to be removed
     * @return true if the listener was removed successfully
     */
    fun removePlaybackStateListener(listener: PlaybackStateListener?): Boolean {
        return listeners.remove(listener)
    }


    override fun logTag(): String {
        return "MY_PLYBK_ST_CLLBK"
    }
}